src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Serialization/ObjectWriter.cs

```diff
@@ -229,27 +254,24 @@ public void WriteValue(object? value)
             {
                 WriteStringValue((string)value);
             }
-            else if (type.IsArray)
-            {
-                var instance = (Array)value;
-
-                if (instance.Rank > 1)
-                {
-                    throw new InvalidOperationException(Resources.Arrays_with_more_than_one_dimension_cannot_be_serialized);
-                }
-
-                WriteArray(instance);
-            }
-            else if (value is Encoding encoding)
-            {
-                WriteEncoding(encoding);
-            }
             else
             {
                 throw new InvalidOperationException($"Unsupported object type: {value.GetType()}");
             }
         }
 
+        public void WriteByteArray(byte[] array)
+        {
+            WriteSpanPreamble<byte>(array.AsSpan(), TypeCode.UInt8);
+            _writer.Write(array);
+        }
+
+        public void WriteCharArray(char[] array)
+        {
+            WriteSpanPreamble<char>(array.AsSpan(), TypeCode.Char);
```

Comment:
>array.AsSpan() [](http://example.com/codeflow?start=36&length=14)

I don't have a strong dislike, but it seems a little odd to do this conversion to a span to send it to a method that doesn't need the span, but just it's length.

Reply:
I'll simplify

Reply:
done.